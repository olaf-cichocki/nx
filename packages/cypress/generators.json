{
  "name": "Nx Cypress",
  "version": "0.1",
  "schematics": {
    "init": {
      "factory": "./src/generators/init/init#cypressInitSchematic",
      "schema": "./src/generators/init/schema.json",
      "description": "Initialize the `@nrwl/cypress` plugin.",
      "aliases": ["ng-add"],
      "hidden": true
    },
    "cypress-project": {
      "factory": "./src/generators/cypress-project/cypress-project#cypressProjectSchematic",
      "schema": "./src/generators/cypress-project/schema.json",
      "description": "Add a Cypress E2E Project"
    },
    "cypress-component-project": {
      "factory": "./src/generators/cypress-component-project/cypress-component-project#cypressComponentProject",
      "schema": "./src/generators/cypress-component-project/schema.json",
      "aliases": ["cy-cmp"],
      "description": "Set up Cypress Component Test for a project"
    },
    "cypress-component-test-file": {
      "factory": "./src/generators/cypress-component-test-file/cypress-component-test-file#cypressComponentTestFiles",
      "schema": "./src/generators/cypress-component-test-file/schema.json",
      "alias": ["cy-test"],
      "description": "Create a Cypress component test for a given component"
    },
    "convert-cypress-ten": {
      "factory": "./src/generators/convert-to-cypress-10/convert-to-cypress-ten#convertCypressProject",
      "schema": "./src/generators/convert-to-cypress-10/schema.json",
      "aliases": ["cy10"],
      "description": "Convert existing Cypress e2e projects to Cypress 10"
    }
  },
  "generators": {
    "init": {
      "factory": "./src/generators/init/init#cypressInitGenerator",
      "schema": "./src/generators/init/schema.json",
      "description": "Initialize the `@nrwl/cypress` plugin.",
      "aliases": ["ng-add"],
      "hidden": true
    },
    "cypress-project": {
      "factory": "./src/generators/cypress-project/cypress-project#cypressProjectGenerator",
      "schema": "./src/generators/cypress-project/schema.json",
      "description": "Add a Cypress E2E Project.",
      "hidden": true
    },
    "cypress-component-project": {
      "factory": "./src/generators/cypress-component-project/cypress-component-project#cypressComponentProject",
      "schema": "./src/generators/cypress-component-project/schema.json",
      "description": "Set up Cypress Component Test for a project",
      "aliases": ["cy-cmp"],
      "hidden": true
    },
    "convert-cypress-ten": {
      "factory": "./src/generators/convert-to-cypress-10/convert-to-cypress-ten#convertCypressProject",
      "schema": "./src/generators/convert-to-cypress-10/schema.json",
      "aliases": ["cy10"],
      "description": "Convert to Cypress 10",
      "hidden": true
    },
    "cypress-component-test-file": {
      "factory": "./src/generators/cypress-component-test-file/cypress-component-test-file#cypressComponentTestFiles",
      "schema": "./src/generators/cypress-component-test-file/schema.json",
      "alias": ["cy-test"],
      "description": "Create a Cypress component test for a given component",
      "hidden": true
    }
  }
}
