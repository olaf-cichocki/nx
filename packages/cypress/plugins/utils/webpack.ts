import { Configuration } from 'webpack';
import { TsconfigPathsPlugin } from 'tsconfig-paths-webpack-plugin';
import { posix } from 'path';
import { getHashDigest, interpolateName } from 'loader-utils';

function getCSSModuleLocalIdent(context, localIdentName, localName, options) {
  // Use the filename or folder name,
  // based on some uses the index.js / index.module.(css|scss|sass) project style
  const fileNameOrFolder = context.resourcePath.match(
    /index\.module\.(css|scss|sass|styl)$/
  )
    ? '[folder]'
    : '[name]';
  // Create a hash based on the file location and class name.
  // Will be unique across a project, and close to globally unique.
  const hash = getHashDigest(
    posix.relative(context.rootContext, context.resourcePath) + localName,
    'md5',
    'base64',
    5
  );
  // Use loaderUtils to find the file or folder name
  const className = interpolateName(
    context,
    `${fileNameOrFolder}_${localName}__${hash}`,
    options
  );
  // Remove the .module that appears in every classname
  // when based on the file and replace all "." with "_".
  return className.replace('.module_', '_').replace(/\./g, '_');
}

const loaderModulesOptions = {
  modules: {
    mode: 'local',
    getLocalIdent: getCSSModuleLocalIdent,
  },
  importLoaders: 1,
};

const commonLoaders = [
  {
    loader: require.resolve('style-loader'),
  },
  {
    loader: require.resolve('css-loader'),
    options: loaderModulesOptions,
  },
];

export const CSS_MODULES_LOADER = {
  // TODO(caleb): does css outside of css-in-js work?
  test: /\.css$|\.scss$|\.sass$|\.less$|\.styl$/,
  oneOf: [
    {
      test: /\.module\.css$/,
      use: commonLoaders,
    },
    {
      test: /\.module\.(scss|sass)$/,
      use: [
        ...commonLoaders,
        {
          loader: require.resolve('sass-loader'),
          options: {
            implementation: require('sass'),
            sassOptions: {
              fiber: false,
              precision: 8,
            },
          },
        },
      ],
    },
    {
      test: /\.module\.less$/,
      use: [
        ...commonLoaders,
        {
          loader: require.resolve('less-loader'),
        },
      ],
    },
    {
      test: /\.module\.styl$/,
      use: [
        ...commonLoaders,
        {
          loader: require.resolve('stylus-loader'),
        },
      ],
    },
  ],
};

export function getLibLoaders(
  compiler: 'babel' | 'swc',
  preset: 'react' | 'next'
) {
  if (compiler === 'swc') {
    return {
      test: /\.([jt])sx?$/,
      loader: require.resolve('swc-loader'),
      exclude: /node_modules/,
      options: {
        jsc: {
          parser: {
            syntax: 'typescript',
            decorators: true,
            tsx: true,
          },
          transform: {
            react: {
              runtime: 'automatic',
            },
          },
          loose: true,
        },
      },
    };
  }

  return {
    test: /\.(js|jsx|mjs|ts|tsx)$/,
    loader: require.resolve('babel-loader'),
    options: {
      presets: [`@nrwl/${preset}/babel`],
      rootMode: 'upward',
      babelrc: true,
    },
  };
}

export function buildBaseWebpackConfig({
  tsConfigPath = 'tsconfig.cy.json',
  compiler = 'babel',
  preset = 'react',
}: {
  tsConfigPath: string;
  compiler: 'swc' | 'babel';
  preset: 'react' | 'next';
}): Configuration {
  const extensions = ['.ts', '.tsx', '.mjs', '.js', '.jsx'];
  return {
    target: 'web',
    resolve: {
      extensions,
      plugins: [
        new TsconfigPathsPlugin({
          configFile: tsConfigPath,
          extensions,
        }) as never,
      ],
    },
    mode: 'development',
    devtool: false,
    output: {
      publicPath: '/',
      chunkFilename: '[name].bundle.js',
    },
    module: {
      rules: [
        {
          test: /\.(bmp|png|jpe?g|gif|webp|avif)$/,
          type: 'asset',
          parser: {
            dataUrlCondition: {
              maxSize: 10_000, // 10 kB
            },
          },
        },
        CSS_MODULES_LOADER,
        getLibLoaders(compiler, preset),
      ],
    },
  };
}
