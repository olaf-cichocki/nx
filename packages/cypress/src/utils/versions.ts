export const nxVersion = '*';
export const eslintPluginCypressVersion = '^2.10.3';
/**
 * TODO(caleb): this is only the mac version, and temp until there is a v10 release
 */
export const cypressVersion =
  'https://cdn.cypress.io/beta/npm/10.0.0/darwin-x64/10.0-release-8765e6481bbc3b8247bd6f64139b8d3676d84c89/cypress.tgz';
export const cypressReactVersion = '^5.12.1';
export const cypressWebpackVersion = '^1.8.0';
export const webpackHttpPluginVersion = '^5.5.0';
export const swcLoaderVersion = '0.1.15';
export const swcCoreVersion = '~1.2.143';
