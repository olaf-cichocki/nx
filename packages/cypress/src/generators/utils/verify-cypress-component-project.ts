import { joinPathFragments, ProjectConfiguration, Tree } from '@nrwl/devkit';
import { CYPRESS_COMPONENT_TEST_TARGET } from '../../utils/project-name';
import { installedCypressVersion } from '../../utils/cypress-version';

export enum ComponentTestingProjectState {
  /**
   * the project already has component testing setup.
   * but did not provide a force flag to override the setup.
   */
  ALREADY_SETUP,
  /**
   * Safe to install cypress.
   */
  INSTALL,
  /**
   * A cypress version > 10 is already installed. No need to install cypress.
   */
  NO_INSTALL,
  /**
   * A cypress version < 10 is installed. Need to upgrade to v10 before allowing a cypress component project to be created.
   */
  UPGRADE,
}

/**
 * Check the installed version of cypress to determine
 * if a project is compatible with cypress component testing.
 */
export function cypressComponentTestingState(
  tree: Tree,
  projectConfig: ProjectConfiguration,
  options: InstallOptions = { force: false }
): ComponentTestingProjectState {
  // if the project already has the component testing target or a cypress.config.ts file then we make sure the --force flag is sent. if not, we throw an error.
  if (
    projectConfig.targets?.[CYPRESS_COMPONENT_TEST_TARGET] ||
    tree.exists(joinPathFragments(projectConfig.root, 'cypress.config.ts'))
  ) {
    if (!options.force) {
      return ComponentTestingProjectState.ALREADY_SETUP;
    }
    return ComponentTestingProjectState.INSTALL;
  }

  const installedVersion = installedCypressVersion();
  // if we don't have a version then it's safe to install cypress
  if (!installedVersion) {
    return ComponentTestingProjectState.INSTALL;
  }

  if (installedVersion >= 10) {
    return ComponentTestingProjectState.NO_INSTALL;
  }

  return ComponentTestingProjectState.UPGRADE;
}

export interface InstallOptions {
  force: boolean;
}
