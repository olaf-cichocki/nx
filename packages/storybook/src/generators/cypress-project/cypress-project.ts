import {
  cypressInitGenerator as _cypressInitGenerator,
  cypressProjectGenerator as _cypressProjectGenerator,
} from '@nrwl/cypress';
import { CypressConfigTransformer } from '@nrwl/cypress/src/utils/config/change-config-transformer';

import {
  getE2eProjectName,
  getUnscopedLibName,
} from '@nrwl/cypress/src/utils/project-name';
import {
  convertNxGenerator,
  formatFiles,
  GeneratorCallback,
  readJson,
  readProjectConfiguration,
  Tree,
  updateJson,
  updateProjectConfiguration,
} from '@nrwl/devkit';
import { Linter } from '@nrwl/linter';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';
import { join } from 'path';
import { safeFileDelete } from '../../utils/utilities';

export interface CypressConfigureSchema {
  name: string;
  js?: boolean;
  directory?: string;
  linter: Linter;
  standaloneConfig?: boolean;
}

export async function cypressProjectGenerator(
  tree: Tree,
  schema: CypressConfigureSchema
) {
  const libConfig = readProjectConfiguration(tree, schema.name);
  const libRoot = libConfig.root;
  const cypressProjectName = `${
    schema.directory ? getUnscopedLibName(libRoot) : schema.name
  }-e2e`;

  const tasks: GeneratorCallback[] = [];

  if (!projectAlreadyHasCypress(tree)) {
    tasks.push(_cypressInitGenerator(tree, {}));
  }

  const installTask = await _cypressProjectGenerator(tree, {
    name: cypressProjectName,
    project: schema.name,
    js: schema.js,
    linter: schema.linter,
    directory: schema.directory,
    standaloneConfig: schema.standaloneConfig,
  });
  tasks.push(installTask);

  const generatedCypressProjectName = getE2eProjectName(
    schema.name,
    libRoot,
    schema.directory
  );
  removeUnneededFiles(tree, generatedCypressProjectName, schema.js);
  addBaseUrlToCypressConfig(tree, generatedCypressProjectName);
  updateAngularJsonBuilder(tree, generatedCypressProjectName, schema.name);

  await formatFiles(tree);

  return runTasksInSerial(...tasks);
}

function removeUnneededFiles(tree: Tree, projectName: string, js: boolean) {
  const { sourceRoot } = readProjectConfiguration(tree, projectName);
  const fileType = js ? 'js' : 'ts';
  safeFileDelete(tree, `${sourceRoot}/integration/app.spec.${fileType}`);
  safeFileDelete(tree, `${sourceRoot}/support/app.po.${fileType}`);
}

function addBaseUrlToCypressConfig(tree: Tree, projectName: string) {
  const projectRoot = readProjectConfiguration(tree, projectName).root;
  const cypressJson = join(projectRoot, 'cypress.json');
  const cypressTs = join(projectRoot, 'cypress.config.ts');
  // TODO(caleb): remove this when cypress < v10 is deprecated

  if (tree.exists(cypressJson)) {
    // cypress < v10
    updateJson(tree, cypressJson, (json) => {
      json.baseUrl = 'http://localhost:4400';
      return json;
    });
  } else if (tree.exists(cypressTs)) {
    // cypress >= v10
    const result = CypressConfigTransformer.addOrUpdateProperties(
      tree.read(cypressTs, 'utf-8'),
      {
        baseUrl: 'http://localhost:4400',
      }
    );
    tree.write(cypressTs, result);
  }
}

function updateAngularJsonBuilder(
  tree: Tree,
  e2eProjectName: string,
  targetProjectName: string
) {
  const project = readProjectConfiguration(tree, e2eProjectName);
  const e2eTarget = project.targets.e2e;
  project.targets.e2e = {
    ...e2eTarget,
    options: <any>{
      ...e2eTarget.options,
      devServerTarget: `${targetProjectName}:storybook`,
    },
    configurations: {
      ci: {
        devServerTarget: `${targetProjectName}:storybook:ci`,
      },
    },
  };
  updateProjectConfiguration(tree, e2eProjectName, project);
}

function projectAlreadyHasCypress(tree: Tree): boolean {
  const packageJsonContents = readJson(tree, 'package.json');
  return (
    (packageJsonContents?.['devDependencies']?.['@nrwl/cypress'] ||
      packageJsonContents?.['dependencies']?.['@nrwl/cypress']) &&
    (packageJsonContents?.['devDependencies']?.['cypress'] ||
      packageJsonContents?.['dependencies']?.['cypress'])
  );
}

export default cypressProjectGenerator;
export const cypressProjectSchematic = convertNxGenerator(
  cypressProjectGenerator
);
