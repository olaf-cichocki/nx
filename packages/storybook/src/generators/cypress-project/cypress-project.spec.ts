import { installedCypressVersion } from '@nrwl/cypress/src/utils/cypress-version';
import { readJson, readProjectConfiguration, Tree } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { Linter } from '@nrwl/linter';
import { libraryGenerator } from '@nrwl/workspace/generators';
import { cypressProjectGenerator } from './cypress-project';

jest.mock('@nrwl/cypress/src/utils/cypress-version');
describe('@nrwl/storybook:cypress-project', () => {
  let tree: Tree;
  let mockedInstalledCypressVersion: jest.Mock<
    ReturnType<typeof installedCypressVersion>
  > = installedCypressVersion as never;

  beforeEach(async () => {
    mockedInstalledCypressVersion.mockReturnValue(10);
    tree = createTreeWithEmptyWorkspace();
    await libraryGenerator(tree, {
      name: 'test-ui-lib',
      standaloneConfig: false,
    });
  });
  afterEach(() => jest.clearAllMocks());

  it('should generate files', async () => {
    await cypressProjectGenerator(tree, {
      name: 'test-ui-lib',
      linter: Linter.EsLint,
      standaloneConfig: false,
    });

    expect(tree.exists('apps/test-ui-lib-e2e/cypress.config.ts')).toBeTruthy();
    const cypressConfig = tree.read(
      'apps/test-ui-lib-e2e/cypress.config.ts',
      'utf-8'
    );
    expect(cypressConfig).toMatchSnapshot();
  });

  it('should update cypress.json file if present', async () => {
    mockedInstalledCypressVersion.mockReturnValue(9);

    await cypressProjectGenerator(tree, {
      name: 'test-ui-lib',
      linter: Linter.EsLint,
      standaloneConfig: false,
    });

    expect(tree.exists('apps/test-ui-lib-e2e/cypress.json')).toBeTruthy();
    const cypressConfig = readJson(tree, 'apps/test-ui-lib-e2e/cypress.json');
    expect(cypressConfig.baseUrl).toEqual('http://localhost:4400');
  });

  it('should update `angular.json` file', async () => {
    await cypressProjectGenerator(tree, {
      name: 'test-ui-lib',
      linter: Linter.EsLint,
      standaloneConfig: false,
    });
    const project = readProjectConfiguration(tree, 'test-ui-lib-e2e');

    expect(project.targets.e2e.options.devServerTarget).toEqual(
      'test-ui-lib:storybook'
    );
    expect(project.targets.e2e.options.headless).toBeUndefined();
    expect(project.targets.e2e.options.watch).toBeUndefined();
    expect(project.targets.e2e.configurations).toEqual({
      ci: { devServerTarget: `test-ui-lib:storybook:ci` },
    });
  });

  it('should generate in the correct folder', async () => {
    await cypressProjectGenerator(tree, {
      name: 'test-ui-lib',
      directory: 'one/two',
      linter: Linter.EsLint,
      standaloneConfig: false,
    });
    const workspace = readJson(tree, 'workspace.json');
    expect(workspace.projects['one-two-test-ui-lib-e2e']).toBeDefined();
    expect(
      tree.exists('apps/one/two/test-ui-lib-e2e/cypress.config.ts')
    ).toBeTruthy();
  });

  it('should make sure the cypress packages are installed', async () => {
    expect(
      readJson(tree, 'package.json').devDependencies['cypress']
    ).toBeFalsy();
    await cypressProjectGenerator(tree, {
      name: 'test-ui-lib',
      directory: 'one/two',
      linter: Linter.EsLint,
      standaloneConfig: false,
    });
    expect(
      readJson(tree, 'package.json').devDependencies['cypress']
    ).toBeTruthy();

    expect(
      readJson(tree, 'package.json').devDependencies['@nrwl/cypress']
    ).toBeTruthy();
  });
});
