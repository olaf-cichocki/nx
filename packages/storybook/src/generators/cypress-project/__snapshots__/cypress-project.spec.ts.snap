// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nrwl/storybook:cypress-project should generate files 1`] = `
"import { defineConfig } from 'cypress';
export default defineConfig({
    baseUrl: 'http://localhost:4400',
    e2e: {
        fileServerFolder: '.',
        fixturesFolder: 'src/fixtures',
        integrationFolder: 'src/e2e',
        supportFile: 'src/support/e2e.ts',
        specPattern: 'src/**/*.cy.{js,ts}',
        pluginsFile: false,
        video: true,
        videosFolder: '../../dist/cypress/apps/test-ui-lib-e2e/videos',
        screenshotsFolder: '../../dist/cypress/apps/test-ui-lib-e2e/screenshots',
        chromeWebSecurity: false
    }
});
"
`;
